sway-core/src/concurrent_slab.rs

```diff
@@ -23,11 +23,15 @@ where
     }
 
     pub fn replace(&self, index: TypeId, prev_value: &T, new_value: T) -> Option<T> {
-        let mut inner = self.inner.write().unwrap();
-        let actual_prev_value = &inner[*index];
-        if actual_prev_value != prev_value {
-            return Some(actual_prev_value.clone());
+        {
+            let read_inner = self.inner.read().unwrap();
+            let actual_prev_value = &read_inner[*index];
+            if actual_prev_value != prev_value {
+                return Some(actual_prev_value.clone());
+            }
         }
```

Comment:
Can we just do this instead?
```suggestion
        let read_inner = self.inner.read().unwrap();
        let actual_prev_value = &read_inner[*index];
        if actual_prev_value != prev_value {
            return Some(actual_prev_value.clone());
        }
```

Reply:
Doesn't look like it, just gave it a try and the test hangs.

Reply:
Because it's waiting for a lock which never comes?  If so this is very fragile and either needs a different approach or lots and lots of comments.

Reply:
The read lock needs to be dropped before a write lock can be given out. The extra scope specifies that the lifetime of the locks allocated here are dropped right after the if statement.  Having a `.read()` and a `.write()` in the same scope sequentially will always deadlock. Scoping the variable down like what @tritao is what I'm used to doing to fix that, because IMO it is prettier than using `drop(lock)`. But yeah some comments wouldn't hurt!

Reply:
Yep exactly, the extra scope is used to limit the lifetime of the lock. I've added some comments to make it a bit clearer.